diff --git a/target/linux/ramips/base-files/etc/board.d/01_leds b/target/linux/ramips/base-files/etc/board.d/01_leds
index 5946f98..5327d00 100755
--- a/target/linux/ramips/base-files/etc/board.d/01_leds
+++ b/target/linux/ramips/base-files/etc/board.d/01_leds
@@ -114,6 +114,10 @@ case $board in
 	esr-9753)
 		set_wifi_led "rt2800pci-phy0::radio"
 		;;
+	ex2700)
+		ucidef_set_led_default "power_r" "POWER (red)" "ex2700:red:power" "0"
+		set_wifi_led "ex2700:green:router"
+		;;
 	f5d8235-v1)
 		set_usb_led "f5d8235-v1:blue:storage"
 		;;
diff --git a/target/linux/ramips/base-files/etc/board.d/02_network b/target/linux/ramips/base-files/etc/board.d/02_network
index d7d016b..a31c508 100755
--- a/target/linux/ramips/base-files/etc/board.d/02_network
+++ b/target/linux/ramips/base-files/etc/board.d/02_network
@@ -247,6 +247,7 @@ ramips_setup_interfaces()
 		ucidef_add_switch_vlan "switch0" "1" "1 2 3 4 5 6t"
 		ucidef_add_switch_vlan "switch0" "2" "0 6t"
 		;;
+	ex2700 | \
 	d105 | \
 	na930 | \
 	omni-emb-hpm|\
diff --git a/target/linux/ramips/base-files/etc/diag.sh b/target/linux/ramips/base-files/etc/diag.sh
index a7e2501..6474ebd 100644
--- a/target/linux/ramips/base-files/etc/diag.sh
+++ b/target/linux/ramips/base-files/etc/diag.sh
@@ -63,6 +63,9 @@ get_status_led() {
 	esr-9753)
 		status_led="esr-9753:orange:power"
 		;;
+	ex2700)
+		status_led="ex2700:red:power"
+		;;
 	f5d8235-v2)
 		status_led="f5d8235v2:blue:router"
 		;;
diff --git a/target/linux/ramips/base-files/lib/ramips.sh b/target/linux/ramips/base-files/lib/ramips.sh
index 4141ddd..8dc05b0 100755
--- a/target/linux/ramips/base-files/lib/ramips.sh
+++ b/target/linux/ramips/base-files/lib/ramips.sh
@@ -223,6 +223,9 @@ ramips_board_detect() {
 	*"Netgear WNCE2001")
 		name="wnce2001"
 		;;
+	*"Netgear EX2700")
+		name="ex2700"
+		;;
 	*"NexAira BC2")
 		name="bc2"
 		;;
diff --git a/target/linux/ramips/base-files/lib/upgrade/platform.sh b/target/linux/ramips/base-files/lib/upgrade/platform.sh
index 2f64d8c..a3d0175 100755
--- a/target/linux/ramips/base-files/lib/upgrade/platform.sh
+++ b/target/linux/ramips/base-files/lib/upgrade/platform.sh
@@ -49,6 +49,7 @@ platform_check_image() {
 	dir-620-d1 | \
 	dir-810l | \
 	e1700 | \
+	ex2700 |\
 	esr-9753 | \
 	f7c027 | \
 	fonera20n | \
diff --git a/target/linux/ramips/dts/EX2700.dts b/target/linux/ramips/dts/EX2700.dts
new file mode 100644
index 0000000..da98bd8
--- /dev/null
+++ b/target/linux/ramips/dts/EX2700.dts
@@ -0,0 +1,148 @@
+/*
+ * Device Tree file for the Netgear EX2700
+ *
+ * Copyright (C) 2016 Joseph C. Lehner <joseph.c.lehner@gmail.com>
+ *
+ * This file is licensed under the terms of the GNU General Public
+ * License version 2.  This program is licensed "as is" without any
+ * warranty of any kind, whether express or implied.
+ */
+
+/dts-v1/;
+
+/include/ "mt7620a.dtsi"
+
+/ {
+	compatible = "ralink,mt7620a-soc";
+	model = "Netgear EX2700";
+
+	chosen {
+		bootargs = "console=ttyS0,57600";
+	};
+
+	palmbus@10000000 {
+
+		gpio0: gpio@600 {
+			status = "okay";
+		};
+
+		gpio1: gpio@638 {
+			status = "okay";
+		};
+
+		spi@b00 {
+			status = "okay";
+
+			m25p80@0 {
+				#address-cells = <1>;
+				#size-cells = <1>;
+				compatible = "mx25l3205d";
+				reg = <0 0>;
+				linux,modalias = "m25p80", "mx25l3205d";
+				spi-max-frequency = <10000000>;
+
+				partition@0 {
+					label = "u-boot";
+					reg = <0x0 0x30000>;
+					read-only;
+				};
+
+				partition@30000 {
+					label = "u-boot-env";
+					reg = <0x30000 0x10000>;
+					read-only;
+				};
+
+				partition@40000 {
+					label = "firmware";
+					reg = <0x40000 0x3b0000>;
+				};
+
+				art: partition@3f0000 {
+					label = "art";
+					reg = <0x3f0000 0x10000>;
+					read-only;
+				};
+			};
+		};
+	};
+
+	ethernet@10100000 {
+		mtd-mac-address = <&art 0x0>;
+	};
+
+	wmac@10180000 {
+		ralink,mtd-eeprom = <&art 0x1000>;
+	};
+
+	pinctrl {
+		state_default: pinctrl0 {
+			default {
+				//  spi refclk: pins 37, 38, 39
+				//       uartf: pins 8, 9, 10, 11, 12, 13, 14
+				//         i2c: pins 1, 2
+				ralink,group = "i2c", "uartf", "spi refclk";
+				ralink,function = "gpio";
+			};
+		};
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+
+		power_g {
+			label = "ex2700:green:power";
+			gpios = <&gpio0 9 1>;
+			default-state = "on";
+		};
+
+		power_r {
+			label = "ex2700:red:power";
+			gpios = <&gpio0 11 1>;
+		};
+
+		device_g {
+			label = "ex2700:green:device";
+			gpios = <&gpio0 13 1>;
+		};
+
+		device_r {
+			label = "ex2700:red:device";
+			gpios = <&gpio0 10 1>;
+		};
+
+		router_g {
+			label = "ex2700:green:router";
+			gpios = <&gpio0 12 1>;
+		};
+
+		router_r {
+			label = "ex2700:red:router";
+			gpios = <&gpio0 14 1>;
+		};
+
+		wps {
+			label = "ex2700:green:wps";
+			gpios = <&gpio1 15 1>;
+		};
+	};
+
+	gpio-keys-polled {
+		compatible = "gpio-keys-polled";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			gpios = <&gpio0 1 1>;
+			linux,code = <0x198>;
+		};
+
+		wps {
+			label = "wps";
+			gpios = <&gpio0 2 1>;
+			linux,code = <0x211>;
+		};
+	};
+};
diff --git a/target/linux/ramips/image/Makefile b/target/linux/ramips/image/Makefile
index 0d762ba..bdb5844 100644
--- a/target/linux/ramips/image/Makefile
+++ b/target/linux/ramips/image/Makefile
@@ -855,6 +855,7 @@ endif
 #
 
 Image/Build/Profile/E1700=$(call BuildFirmware/UMedia/$(1),$(1),e1700,E1700,0x013326)
+Image/Build/Profile/EX2700=$(call BuildFirmware/Default4M/$(1),$(1),ex2700,EX2700)
 Image/Build/Profile/MT7620a=$(call BuildFirmware/Default8M/$(1),$(1),mt7620a,MT7620a)
 Image/Build/Profile/MT7620a_MT7610e=$(call BuildFirmware/Default8M/$(1),$(1),mt7620a_mt7610e,MT7620a_MT7610e)
 Image/Build/Profile/MT7620a_MT7530=$(call BuildFirmware/Default8M/$(1),$(1),mt7620a_mt7530,MT7620a_MT7530)
@@ -893,6 +894,7 @@ Image/Build/Profile/MicroWRT=$(call BuildFirmware/CustomFlash/$(1),$(1),microwrt
 ifeq ($(SUBTARGET),mt7620)
 define Image/Build/Profile/Default
 	$(call Image/Build/Profile/E1700,$(1))
+	$(call Image/Build/Profile/EX2700,$(1))
 	$(call Image/Build/Profile/MT7620a,$(1))
 	$(call Image/Build/Profile/MT7620a_MT7610e,$(1))
 	$(call Image/Build/Profile/MT7620a_MT7530,$(1))
diff --git a/target/linux/ramips/mt7620/profiles/netgear.mk b/target/linux/ramips/mt7620/profiles/netgear.mk
new file mode 100644
index 0000000..d1b073a
--- /dev/null
+++ b/target/linux/ramips/mt7620/profiles/netgear.mk
@@ -0,0 +1,16 @@
+#
+# Copyright (C) 2016 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+define Profile/EX2700
+	NAME:=Netgear EX2700
+	PACKAGES:=-kmod-usb-core -kmod-usb2 -kmod-usb-ohci -kmod-ledtrig-usbdev
+endef
+
+define Profile/EX2700/Description
+	Support for Netgear EX2700
+endef
+$(eval $(call Profile,EX2700))
diff --git a/target/linux/ramips/patches-3.18/0303-fix-mt7530-vlan.patch b/target/linux/ramips/patches-3.18/0303-fix-mt7530-vlan.patch
new file mode 100644
index 0000000..d38dddf
--- /dev/null
+++ b/target/linux/ramips/patches-3.18/0303-fix-mt7530-vlan.patch
@@ -0,0 +1,25 @@
+--- a/drivers/net/ethernet/ralink/mt7530.c
++++ b/drivers/net/ethernet/ralink/mt7530.c
+@@ -730,7 +730,6 @@ mt7530_probe(struct device *dev, void __
+ 
+ 	mt7530->base = base;
+ 	mt7530->bus = bus;
+-	mt7530->global_vlan_enable = vlan;
+ 
+ 	swdev = &mt7530->swdev;
+ 	if (bus) {
+@@ -759,8 +758,13 @@ mt7530_probe(struct device *dev, void __
+ 
+ 
+ 	map = mt7530_find_mapping(dev->of_node);
+-	if (map)
++	if (map) {
++		mt7530->global_vlan_enable = vlan;
+ 		mt7530_apply_mapping(mt7530, map);
++	} else {
++		mt7530->global_vlan_enable = 0;
++		dev_info(dev, "no ralink,port-map specified - disabling VLAN\n");
++	}
+ 	mt7530_apply_config(swdev);
+ 
+ 	/* magic vodoo */
